editor/editor_node.cpp

```diff
@@ -1724,6 +1724,131 @@ void EditorNode::save_resource_as(const Ref<Resource> &p_resource, const String
 	file->popup_file_dialog();
 }
 
+void EditorNode::gather_resources(const Variant &p_variant, List<Ref<Resource>> &p_l, bool subresources, bool allow_external) {
+	Variant::Type type = p_variant.get_type();
+	if (p_variant.get_validated_object() == nullptr && type == Variant::OBJECT) {
+		return;
+	}
+
+	if (type != Variant::OBJECT && type != Variant::ARRAY && type != Variant::DICTIONARY) {
+		return;
+	}
+
+	if (type == Variant::ARRAY) {
+		Array arr = p_variant;
+		for (int i = 0; i < arr.size(); i++) {
+			Ref<Resource> res = arr[i];
+			if (res.is_valid()) {
+				p_l.push_back(res);
+			}
+			gather_resources(arr[i], p_l, subresources, allow_external);
+		}
+		return;
+	}
+
+	if (type == Variant::DICTIONARY) {
+		Dictionary dict = p_variant;
+		for (const KeyValue<Variant, Variant> &kv : dict) {
+			Ref<Resource> res_key = kv.key;
+			Ref<Resource> res_value = kv.value;
+			if (res_key.is_valid()) {
+				p_l.push_back(res_key);
+			}
+			if (res_value.is_valid()) {
+				p_l.push_back(res_value);
+			}
+			gather_resources(kv.key, p_l, subresources, allow_external);
+			gather_resources(kv.value, p_l, subresources, allow_external);
+		}
+		return;
+	}
+
+	List<PropertyInfo> pinfo;
+	p_variant.get_property_list(&pinfo);
+
+	for (const PropertyInfo &E : pinfo) {
+		if (!(E.usage & PROPERTY_USAGE_EDITOR) || E.name == "script") {
+			continue;
+		}
+
+		Variant property_value = p_variant.get(E.name);
+		Variant::Type property_type = property_value.get_type();
+
+		if (property_type == Variant::ARRAY || property_type == Variant::DICTIONARY) {
+			gather_resources(property_value, p_l, subresources, allow_external);
+		}
+
+		Ref<Resource> res = property_value;
+		if (res.is_null()) {
+			continue;
+		}
+		if (!allow_external) {
+			if (!res->is_built_in() && res->get_path().get_slice("::", 0) != get_edited_scene()->get_scene_file_path()) {
+				if (!res->get_path().is_empty()) {
+					continue;
+				}
+			}
+		}
+		p_l.push_back(res.ptr());
+		if (subresources) {
+			gather_resources(res.ptr(), p_l, subresources, allow_external);
+		}
+	}
+}
+
+void EditorNode::update_resource_count(Node *p_node, bool remove) {
+	if (!get_edited_scene()) {
+		return;
+	}
+
+	List<Ref<Resource>> res_list;
+	gather_resources(p_node, res_list, true);
+
+	for (Ref<Resource> R : res_list) {
+		Resource *res = R.ptr();
+		if (resource_count[res].find(p_node)) {
+			if (remove) {
+				resource_count[res].erase(p_node);
+			}
+		} else {
+			resource_count[res].push_back(p_node);
+		}
+	}
+
+	emit_signal(SNAME("resource_counter_changed"));
+}
+
+int EditorNode::get_resource_count(Ref<Resource> p_res) const {
+	return (resource_count.has(p_res.ptr())) ? resource_count[p_res.ptr()].size() : 0;
```

Comment:
Avoid double lookups
```suggestion
	List<Node *> *L = resource_count.getptr(p_res.ptr());
	return L ? L->size() : 0;
```
Same below.